### PR TITLE
Introduce interfaces to allow customisation of NuCache

### PR DESCRIPTION
**Pull request by nzdev**
### Prerequisites

- [ ] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes 

### Description

What?
Allow replacement of BPlusTree implementation, Serializers for BPlusTree, ContentData.

Why?
Out of the box works well but can be optimised further in specific scenarios using alternative serialization libraries and techniques.

How:
Introduce ITransactableDictionary interface for BPlusTree.
Introduce ITransactableDictionaryFactory interface for creating ITransactableDictionary.
Inject ISerializer<IContentNodeKit> to BPlussTreeTransactableDictionaryFactory to allow full control over serialization to and from BPlusTree.
Introduce IContentData for ContentData.

Testing:
Replace any of the interfaces mentioned above using a composer.

Replaces PR 8781

### CODE REVIEW

File: src/Umbraco.Web/PublishedCache/NuCache/PublishedSnapshotService.cs [hunk 1]
```diff
@@ -172,18 +171,11 @@ internal class PublishedSnapshotService : PublishedSnapshotServiceBase
         /// </remarks>
         private void MainDomRegister()
         {
-            var path = GetLocalFilesPath();
-            var localContentDbPath = Path.Combine(path, "NuCache.Content.db");
-            var localMediaDbPath = Path.Combine(path, "NuCache.Media.db");
+            // if both local databases exist then Get will open them, else new databases will be created
+            _localContentDb = _transactableDictionaryFactory.Get(ContentCacheEntityType.Document);
+            _localMediaDb = _transactableDictionaryFactory.Get(ContentCacheEntityType.Document);
```
Comment (nzdev) [inline src/Umbraco.Web/PublishedCache/NuCache/PublishedSnapshotService.cs:176]: ContentCacheEntityType.Media

File: src/Umbraco.Web/PublishedCache/NuCache/ITransactable.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Umbraco.Web.PublishedCache.NuCache
+{
+    public interface ITransactable : IDisposable
```
Comment (Shazwazza) [inline src/Umbraco.Web/PublishedCache/NuCache/ITransactable.cs:9]: Do we need this? Not sure that this specific interface is actually used?

File: src/Umbraco.Web/PublishedCache/NuCache/ITransactableDictionaryFactory.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        /// </summary>
+        /// <param name="errors">The errors, if any.</param>
+        /// <returns>A value indicating whether the ITransactableDictionaryFactory has the proper environment to run.</returns>
+        bool EnsureEnvironment(out IEnumerable<string> errors);
+
+        /// <summary>
+        /// Whether the dictionary has been populated
+        /// </summary>
+        /// <returns></returns>
+        bool IsPopulated(ContentCacheEntityType entityType);
```
Comment (Shazwazza) [inline src/Umbraco.Web/PublishedCache/NuCache/ITransactableDictionaryFactory.cs:24]: What is the reason for having IsPopulated on both the ITransactableDictionaryFactory and the ITransactableDictionary? I can only see the factory usage like `if ((_localContentDb != null && _localContentDb.IsPopulated()) || _transactableDictionaryFactory.IsPopulated(ContentCacheEntityType.Document))` but I'm unsure why?

**Review by Shazwazza — COMMENTED** [review]
Left a couple comments.

Also when you say

> Introduce IContentData for ContentData.

That doesn't seem to be the case, maybe the comment is left over from some refactoring?

(These changes are going to be slightly nightmarish to port to netcore since a lot has changed there now :P ... but we'll get there)

**Comment by nzdev** [comment]
Replaced by 9857

**Pull request closed without merging**